src/ui/marker.js

```diff
@@ -1,36 +1,71 @@
 // @flow
 
 const DOM = require('../util/dom');
+const util = require('../util/util');
+const {bindAll} = require('../util/util');
 const LngLat = require('../geo/lng_lat');
 const Point = require('@mapbox/point-geometry');
 const smartWrap = require('../util/smart_wrap');
-const {bindAll} = require('../util/util');
 
 import type Map from './map';
 import type Popup from './popup';
 import type {LngLatLike} from "../geo/lng_lat";
 import type {MapMouseEvent} from './events';
 
+const defaultOptions = {
+    anchor: 'middle',
+    offset: [0, 0]
+};
+
+export type Anchor = 'middle' | 'top' | 'bottom' | 'left' | 'right' | 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
+export type Offset = number | PointLike | {[Anchor]: PointLike};
+
+export type MarkerOptions = {
+    anchor: Anchor,
+    offset: Offset
+}
+
 /**
  * Creates a marker component
- * @param element DOM element to use as a marker. If left unspecified a default SVG will be created as the DOM element to use.
- * @param options
- * @param options.offset The offset in pixels as a {@link PointLike} object to apply relative to the element's center. Negatives indicate left and up.
+ * @param {Object} [element] DOM element to use as a marker. If left unspecified a default SVG will be created as the DOM element to use.
+ * @param {Object} [options]
+ * @param {string} [options.anchor] - A string indicating the markers's location relative to
+ *   the coordinate set via {@link Marker#setLngLat}.
+ *   Options are `'middle'`, `'top'`, `'bottom'`, `'left'`, `'right'`, `'top-left'`, `'top-right'`, `'bottom-left'`, and `'bottom-right'`.
+ *   If unset the anchor will be dynamically set to ensure the marker falls within the map container with a preference for `'middle'` by default
+ * @param {number|PointLike|Object} [options.offset] The offset in pixels as a {@link PointLike} object to apply relative to the element's anchor.
  * @example
+ * var markerHeight = 50, markerRadius = 10, linearOffset = 25;
```

Comment:
Even if it ends up supporting the same dynamic anchor logic as Popup, I don't think such a complicated use case should be the default example.

Reply:
this is a good point, i'll make an example based on defaults